### PR TITLE
Add DNA Foundation Models

### PR DESCRIPTION
**Pull request by aerdasaliko**
This PR adds a new Galaxy tool, **DNA Foundation Models**, integrating the DNABERT-2 foundation model into Galaxy.

The tool supports two main execution modes:

* **Embedding mode**: generates sequence embeddings from FASTA input
* **Fine-tuning mode**: trains or adapts models using tabular datasets, with support for both full fine-tuning and LoRA-based parameter-efficient tuning

In embedding mode, the tool processes a FASTA file by batching sequences through the model to generate embeddings. Pooling is configurable (`CLS`, mean, or max), and outputs are written as a tabular file. Optionally, attention visualizations can be enabled, producing additional outputs via BertViz and related utilities.

In fine-tuning mode, the tool takes tabular datasets (train/dev/test) in `(sequence, label)` format and passes training parameters directly to the underlying training script. It exposes the main Trainer configuration options, such as batch size, learning rate, number of epochs, evaluation steps, and logging frequency.

Two training strategies are supported:

* Full fine-tuning
* LoRA-based fine-tuning (optionally enabled, with configurable rank, alpha, dropout, and target modules)

Training artifacts (evaluation metrics, predictions, trainer state) are returned as outputs. Optionally, the trained model is returned as a `.tar.gz` archive.

### Notes

The original DNABERT-2 model is based on a MosaicBERT architecture that relies on a version of Flash Attention not compatible with newer PyTorch releases. This tool instead uses model weights loaded through an updated MosaicBERT implementation to ensure compatibility.

The model can be retrieved using the following script:
[model-retrieval.py](https://github.com/user-attachments/files/31384995/model-retrieval.py)

FOR CONTRIBUTOR:
* [x] I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] License permits unrestricted use (educational + commercial)
* [x] This PR adds a new tool or tool collection
* [ ] This PR updates an existing tool or tool collection
* [ ] This PR does something else (explain below)

There are two labels that allow to ignore specific (false positive) tool linter errors:

* `skip-version-check`: Use it if only a subset of the tools has been updated in a suite.
* `skip-url-check`: Use it if github CI sees 403 errors, but the URLs work.

### CODE REVIEW

**Comment by anuprulez** [comment]
thanks @aerdasaliko for the work!

File: tools/dna_foundation_models/tool-data/huggingface.loc [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+# Administrators: copy this file to your Galaxy tool-data directory as
+# "huggingface.loc" and update each path to point to the actual model files.
+# Models can be downloaded from: ..... !!!!!!!!!!!!!!!!!!!!
+#
+# For the full shared schema reference see:
+#   https://galaxyproject.org/news/2026-04-13-huggingface-data-table/
+#
+#
+# Classification models
+local-dnabert2-model-1	DNABERT-2-117M	fill-mask	text	dnabert2	1	"path-to-model"
```
Comment (anuprulez) [inline tools/dna_foundation_models/tool-data/huggingface.loc]:
```suggestion
DNABERT-2-117M	DNABERT-2-117M	feature-extraction	dna-sequence	dnabert2	1	"path-to-model"
```

Comment (anuprulez) [reply tools/dna_foundation_models/tool-data/huggingface.loc]: It seems none of the `pipeline_tag` values from HF is suitable for DNABERT-2 model.

File: tools/dna_foundation_models/dna_foundation_models.xml [hunk 2]
```diff
@@ -0,0 +1,402 @@
+<tool id="dna_foundation_models" name="DNA Foundation Models" version="0.1.0+galaxy0" profile="21.05">
```
Comment (anuprulez) [inline tools/dna_foundation_models/dna_foundation_models.xml]:
```suggestion
<tool id="dna_foundation_models" name="DNA Foundation Models" version="0.1.0+galaxy0" profile="24.2">
```

File: tools/dna_foundation_models/dna_foundation_models.xml [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+
+                <section name="training_params">
+                    <param name="model_max_length" value="128"/>
+                    <param name="batch_train" value="8"/>
+                    <param name="batch_eval" value="8"/>
+                    <param name="grad_acc" value="1"/>
+                    <param name="lr" value="5e-5"/>
+                    <param name="weight_decay" value="0.01"/>
+                    <param name="epochs" value="1"/>
+                    <param name="save_steps" value="50"/>
```
Comment (anuprulez) [inline tools/dna_foundation_models/dna_foundation_models.xml]: `save_steps` is not defined in the tool's UI

File: tools/dna_foundation_models/dna_foundation_models.xml [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+            </conditional>
+
+            <assert_stderr>
+                <has_text_matching expression="Parameter 'selected_model': requires a value, but no legal values defined"/>
+            </assert_stderr>
+        </test>
+
+    </tests>
+
+    <help>
```
Comment (anuprulez) [inline tools/dna_foundation_models/dna_foundation_models.xml:315]: `help` section is not formatted .. `planemo lint *.xml` gives errors.

Comment (aerdasaliko) [reply tools/dna_foundation_models/dna_foundation_models.xml:315]: Resolved in https://github.com/bgruening/galaxytools/pull/1965/commits/8ead87295635d865462c0670a3b14f35c790d2e6

**Comment by anuprulez** [comment]
@aerdasaliko thanks for the updates

I think tool linting tests are failing -> recommendations are present to fix these (For example - `'import numpy' should be before 'import torch'`)

```
tools/dna_foundation_models/embed_sequence.py:3:1: I201 Missing newline between import groups. 'import torch' is identified as Third Party and 'import csv' is identified as Stdlib.
tools/dna_foundation_models/embed_sequence.py:4:1: I100 Import statements are in the wrong order. 'import numpy' should be before 'import torch'
tools/dna_foundation_models/embed_sequence.py:5:1: I101 Imported names are in the wrong order. Should be AutoConfig, AutoModel, AutoTokenizer
tools/dna_foundation_models/finetune.py:2:1: I100 Import statements are in the wrong order. 'import csv' should be before 'import os'
tools/dna_foundation_models/finetune.py:6:1: I101 Imported names are in the wrong order. Should be Any, Dict, List, Optional, Sequence, Tuple, Union
tools/dna_foundation_models/finetune.py:6:1: F401 'typing.List' imported but unused
tools/dna_foundation_models/finetune.py:10:1: I100 Import statements are in the wrong order. 'import sklearn' should be before 'import transformers'
tools/dna_foundation_models/finetune.py:11:1: I100 Import statements are in the wrong order. 'import numpy' should be before 'import sklearn'
tools/dna_foundation_models/finetune.py:13:1: I100 Import statements are in the wrong order. 'from scipy.stats import pearsonr, spearmanr' should be before 'from torch.utils.data import Dataset'
tools/dna_foundation_models/finetune.py:15:1: I101 Imported names are in the wrong order. Should be get_peft_model, LoraConfig
tools/dna_foundation_models/finetune.py:15:1: I100 Import statements are in the wrong order. 'from peft import LoraConfig, get_peft_model' should be before 'from scipy.stats import pearsonr, spearmanr'
tools/dna_foundation_models/finetune.py:15:1: I202 Additional newline in a group of imports. 'from peft import LoraConfig, get_peft_model' is identified as Third Party and 'from scipy.stats import pearsonr, spearmanr' is identified as Third Party.
tools/dna_foundation_models/visualize_attn.py:7:1: I100 Import statements are in the wrong order. 'import plotly.express' should be before 'import plotly.graph_objects'
tools/dna_foundation_models/visualize_attn.py:7:1: F401 'plotly.express as px' imported but unused
tools/dna_foundation_models/visualize_attn.py:9:1: I100 Import statements are in the wrong order. 'from bertviz import head_view, model_view' should be before 'from plotly.subplots import make_subplots'
```

File: tools/dna_foundation_models/dna_foundation_models.xml [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+        <requirement type="package" version="0.19.1">peft</requirement>
+        <requirement type="package" version="2.3.0">omegaconf</requirement>
+        <requirement type="package" version="1.7.2">scikit-learn</requirement>
+        <requirement type="package" version="1.13.0">accelerate</requirement>
+        <requirement type="package" version="0.4.6">evaluate</requirement>
+        <requirement type="package" version="6.9.0">plotly</requirement>
+        <requirement type="package" version="3.1.5">openpyxl</requirement>
+        <requirement type="package" version="1.4.1">bertviz</requirement>
+        <requirement type="package" version="2.3.3">pandas</requirement>
+        <requirement type="package" version="2.7.4.post1">flash-attn</requirement>
```
Comment (anuprulez) [inline tools/dna_foundation_models/dna_foundation_models.xml]:
```suggestion
        <requirement type="package" version="2.7.4">flash-attn</requirement>
```

Comment (anuprulez) [reply tools/dna_foundation_models/dna_foundation_models.xml]: conda is not able to resolve with this package version `2.7.4.post1`

File: tools/dna_foundation_models/dna_foundation_models.xml [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+                       help="FASTA file containing the DNA sequences to embed. Each FASTA record is processed 
+                       as one sequence."/>
+                <param name="pooling" type="select" label="Pooling strategy"
+                       help="Select how token-level representations are combined into one embedding vector 
+                       for each input sequence.">
+                    <option value="cls" selected="true">CLS</option>
+                    <option value="mean">mean</option>
+                    <option value="max">maximum</option>
+                </param>
+                <param name="batch_size" type="integer" value="4" min="4" max="512"
```
Comment (anuprulez) [inline tools/dna_foundation_models/dna_foundation_models.xml]:
```suggestion
                <param name="batch_size" type="integer" value="4" min="1" max="512"
```

File: tools/dna_foundation_models/dna_foundation_models.xml (hunk 3 above)
Comment (aerdasaliko) [reply tools/dna_foundation_models/dna_foundation_models.xml]: I removed it from the UI, since the checkpoints won't be visible to the user. It's now also removed from the tests

File: tools/dna_foundation_models/dna_foundation_models.xml [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+                <param name="batch_size" value="4"/>
+            </conditional>
+
+            <assert_stderr>
+                <has_text_matching expression="Parameter 'selected_model': requires a value, but no legal values defined"/>
+            </assert_stderr>
+        </test>
+
+        <!-- Fine-tune: unknown model path -->
+        <test expect_failure="true" expect_exit_code="5">
```
Comment (anuprulez) [inline tools/dna_foundation_models/dna_foundation_models.xml:282]: please increase the test coverage. Add tests for other modes of the tool - embedding prediction, fine-tuning for regression task etc.

File: tools/dna_foundation_models/dna_foundation_models.xml [hunk 8]
```diff
@@ -0,0 +1,441 @@
...
+                        --attn_ids all
+                    #end if
+                    #if $tasks.lora_section.use_lora == 'true'
+                        --use_lora
+                        --lora_r '$tasks.lora_section.lora_r'
+                        --lora_alpha '$tasks.lora_section.lora_alpha'
+                        --lora_dropout '$tasks.lora_section.lora_dropout'
+                        --lora_target_modules '$tasks.lora_section.target_modules'
+                    #end if
+                ;
```
Comment (anuprulez) [inline tools/dna_foundation_models/dna_foundation_models.xml:68]: is this ";" necessary?

File: tools/dna_foundation_models/dna_foundation_models.xml [hunk 9]
```diff
@@ -0,0 +1,441 @@
...
+                    --pooling '$tasks.pooling'
+                    --output results/embeddings.tsv
+                #if $attn_vis == 'true'
+                    --attn_vis
+                    --viz_dir results/attn_vis
+                    --attn_ids all
+                #end if
+                ;
+            #else
+                echo 'Unknown task: $tasks.selected_task'
```
Comment (anuprulez) [inline tools/dna_foundation_models/dna_foundation_models.xml:88]: do we have any task other than `finetune` and `embed` defined? The `#else` condition may not be necessary.

File: tools/dna_foundation_models/dna_foundation_models.xml [hunk 10]
```diff
@@ -0,0 +1,441 @@
...
+                <column name="value" index="6"/>
+                <filter type="static_value" value="dnabert2" column="4"/>
+                <filter type="static_value" value="1" column="5"/>
+            </options>
+            <validator type="no_options" message="No pretrained DNABERT models are installed on this Galaxy instance."/>
+        </param>
+
+        <param name="attn_vis" type="boolean" truevalue="true" falsevalue="false" 
+            checked="false" label="Visualize attention"
+            help="Generate attention visualizations for all sequences in the input file (classification)/test file (regression)."/>
```
Comment (anuprulez) [inline tools/dna_foundation_models/dna_foundation_models.xml:107]: I am not sure if I understood it `input file (classification)/test file (regression)`?

File: tools/dna_foundation_models/dna_foundation_models.xml [hunk 11]
```diff
@@ -0,0 +1,441 @@
...
+                       help="Tabular file used to train the model. Must contain two columns: sequence, label."/>
+                <param name="dev_dataset" type="data" format="tabular" label="Validation dataset"
+                       help="Tabular file used to evaluate the model during training. Same format as the training 
+                       dataset."/>
+                <param name="test_dataset" type="data" format="tabular" label="Test dataset"
+                       help="Tabular file used for final evaluation after training. Same format as training data. 
+                            This dataset should not be used during training to ensure unbiased performance estimates."/>
+
+                <section name="training_params" title="Training parameters" expanded="true">
+                    <param name="model_max_length" type="integer" value="128" min="16" max="4096"
```
Comment (anuprulez) [inline tools/dna_foundation_models/dna_foundation_models.xml:174]: `4096` is the maximum value set by the model?

File: tools/dna_foundation_models/finetune.py [hunk 12]
```diff
@@ -0,0 +1,443 @@
...
+    )
+
+    use_lora: bool = field(
+        default=False,
+        metadata={"help": "Whether to enable LoRA fine-tuning"}
+    )
+
+    # LoRA args
+    lora_r: int = field(default=8, metadata={"help": "hidden dimension for LoRA"})
+    lora_alpha: int = field(default=32, metadata={"help": "alpha for LoRA"})
```
Comment (anuprulez) [inline tools/dna_foundation_models/finetune.py:38]: for `lora_alpha`, default is 32 but in the XML it has a default value of 16. We should use the same default at both places. Also, please check this for other parameters.

```
<param name="lora_alpha" type="integer" value="16" label="LoRA alpha" min="1" max="512"
                               help="Scaling factor applied to the LoRA updates. The magnitude of the adaptation's 
                               contribution to the model's output is scaled by alpha/r."/>
```

File: tools/dna_foundation_models/finetune.py [hunk 13]
```diff
@@ -0,0 +1,443 @@
...
+    }
+    return results
+
+
+def calculate_regression_metrics(predictions: np.ndarray, labels: np.ndarray):
+    valid_mask = labels != -100
+    preds = predictions[valid_mask]
+    targets = labels[valid_mask]
+
+    mse = sklearn.metrics.mean_squared_error(targets, preds)
```
Comment (anuprulez) [inline tools/dna_foundation_models/finetune.py:176]: please import `sklearn.metrics` and then use it at other places

File: tools/dna_foundation_models/finetune.py [hunk 14]
```diff
@@ -0,0 +1,443 @@
...
+        if problem_type == "classification":
+            probs = torch.softmax(torch.tensor(logits), dim=-1).numpy()
+            predictions = np.argmax(probs, axis=-1)
+            results = calculate_classification_metrics(predictions, labels)
+
+            valid_mask = labels != -100
+            valid_labels = labels[valid_mask]
+
+            if probs.shape[-1] == 2 and len(np.unique(valid_labels)) > 1:  # for binary classification
+                results["roc_auc"] = sklearn.metrics.roc_auc_score(valid_labels, probs[valid_mask, 1])
```
Comment (anuprulez) [inline tools/dna_foundation_models/finetune.py:237]: Same comment for `sklearn.metrics`

**Review by anuprulez — COMMENTED** [review]
@aerdasaliko I have added a few more comments.

**Comment by anuprulez** [comment]
i have checked the features, they work fine. But, there are some cosmetic changes in the code that should be fixed.
